File: examples/python/activity_recording.py
#!/usr/bin/env python3
"""Model Health Python SDK — capture workflow demo.

Walks through the full capture workflow:
  1. Create a session (QR code saved locally for pairing)
  2. Select an existing subject or create a new one
  3. Calibrate cameras using a checkerboard pattern
  4. Calibrate the subject (neutral standing pose)
  5. Record a movement activity
  6. Wait for activity upload and processing

Requires cameras to be connected and ready via the Model Health companion iOS app.

Usage:
    activity_recording.py [<api_key>]
"""

import os
import sys
import time
import urllib.request

from docopt import docopt

from modelhealth import (
    ActivityStatus,
    ActivityStatusUploading,
    CalibrationStatus,
    CalibrationStatusProcessing,
    CalibrationStatusUploading,
    CheckerboardDetails,
    CheckerboardPlacement,
    ModelHealthError,
    ModelHealthService,
    SubjectParameters,
)
from _prompts import confirm, pick_one
from _utils import DOWNLOADS_DIR, load_api_key

# ---------------------------------------------------------------------------
# Checkerboard presets
# ---------------------------------------------------------------------------

_CHECKERBOARD_PRESETS = [
    (4, 5, 35, "4 × 5  —  35 mm squares  (standard A4)"),
    (4, 5, 50, "4 × 5  —  50 mm squares  (large A3)"),
    (None, None, None, "Other — enter manually"),
]

# ---------------------------------------------------------------------------
# Calibration callback
# ---------------------------------------------------------------------------

_CB_WIDTH = 48


def _calibration_callback(status):
    if isinstance(status, CalibrationStatusUploading):
        msg = f"  Uploading ({status.uploaded}/{status.total} cameras)..."
        print(msg.ljust(_CB_WIDTH), end="\r", flush=True)
    elif isinstance(status, CalibrationStatusProcessing):
        pct = f"{status.percent}%" if status.percent is not None else "--%"
        msg = f"  Processing ({pct})..."
        print(msg.ljust(_CB_WIDTH), end="\r", flush=True)
    elif status == CalibrationStatus.recording:
        print("  Recording...".ljust(_CB_WIDTH), end="\r", flush=True)
    elif status == CalibrationStatus.done:
        print("  Done.".ljust(_CB_WIDTH))


# ---------------------------------------------------------------------------
# Activity polling helper
# ---------------------------------------------------------------------------

def _poll_activity(service, activity, interval=5):
    """Block until the activity finishes uploading and processing."""
    while True:
        status = service.activity_status(activity)
        if isinstance(status, ActivityStatusUploading):
            print(
                f"  Uploading ({status.uploaded}/{status.total} cameras)...  ",
                end="\r", flush=True,
            )
        elif status == ActivityStatus.processing:
            print("  Processing...                              ", end="\r", flush=True)
        else:
            print()
            return status
        time.sleep(interval)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main(api_key):
    print("Connecting...")
    try:
        service = ModelHealthService(api_key)
    except ModelHealthError as exc:
        sys.exit(f"Failed to initialise: {exc}")

    # Session
    print("\nCreating session...")
    try:
        session = service.create_session()
    except ModelHealthError as exc:
        sys.exit(f"Failed to create session: {exc}")
    print(f"  Session ID: {session.id}")

    if not session.qrcode:
        sys.exit("Session has no QR code — cannot pair cameras.")

    os.makedirs(DOWNLOADS_DIR, exist_ok=True)
    qr_path = os.path.join(DOWNLOADS_DIR, "qr-code.png")
    try:
        urllib.request.urlretrieve(session.qrcode, qr_path)
        print(f"  QR code saved to: {qr_path}")
    except Exception as exc:
        sys.exit(f"Failed to download QR code: {exc}")

    print("  Pair your cameras using the Model Health companion iOS app before continuing.")
    input("\nPress Enter when cameras are ready...")

    # Camera calibration
    print("\nCheckerboard configuration:\n")
    preset = pick_one(
        _CHECKERBOARD_PRESETS,
        "Select checkerboard",
        lambda p: p[3],
    )
    rows, columns, square_size, _ = preset

    if rows is None:
        rows        = int(input("  Internal rows: ").strip())
        columns     = int(input("  Internal columns: ").strip())
        square_size = int(input("  Square size (mm): ").strip())

    print("\nCheckerboard placement:\n")
    placement_value, _ = pick_one(
        [
            (CheckerboardPlacement.perpendicular, "Perpendicular (upright, facing cameras)"),
            (CheckerboardPlacement.parallel,      "Parallel (flat on the floor)"),
        ],
        "Select placement",
        lambda p: p[1],
    )

    checkerboard = CheckerboardDetails(
        rows=rows,
        columns=columns,
        square_size=square_size,
        placement=placement_value,
    )

    input("\nPress Enter to start camera calibration...")
    print("Calibrating cameras...")
    try:
        service.calibrate_camera(session, checkerboard, _calibration_callback)
    except ModelHealthError as exc:
        sys.exit(f"Camera calibration failed: {exc}")
    print("Camera calibration complete.")

    # Subject
    print("\nFetching subjects...")
    try:
        subjects = service.subject_list()
    except ModelHealthError as exc:
        sys.exit(f"Failed to fetch subjects: {exc}")

    if subjects and confirm(f"Found {len(subjects)} subject(s). Select an existing one?", default=True):
        print()
        subject = pick_one(
            subjects,
            "Select subject",
            lambda s: f"{s.name}  (ID {s.id})",
        )
        print(f"  Using: {subject.name}")
    else:
        print("\nNew subject details:")
        name   = input("  Name: ").strip() or "Anonymous"
        weight = float(input("  Weight (kg): ").strip())
        height = float(input("  Height (cm): ").strip())

        params = SubjectParameters(name=name, weight=weight, height=height)
        print("Creating subject...")
        try:
            subject = service.create_subject(params)
        except ModelHealthError as exc:
            sys.exit(f"Failed to create subject: {exc}")
        print(f"  Subject created: {subject.name} (ID {subject.id})")

    # Subject calibration
    input(f"\nAsk {subject.name} to stand in the neutral pose, then press Enter...")
    print("Calibrating subject...")
    try:
        service.calibrate_subject(subject, session, _calibration_callback)
    except ModelHealthError as exc:
        sys.exit(f"Subject calibration failed: {exc}")
    print("Subject calibration complete.")

    # Recording
    activity_name = input("\nActivity name (e.g. cmj, squat): ").strip() or "activity"

    input(f"\nAsk {subject.name} to get ready, then press Enter to start recording...")
    print("Recording...")
    try:
        activity = service.start_recording(activity_name, session)
    except ModelHealthError as exc:
        sys.exit(f"Failed to start recording: {exc}")
    print(f"  Recording started (activity {activity.id}).")

    input("\nPress Enter when the movement is complete to stop recording...")
    print("Stopping recording...")
    try:
        service.stop_recording(session)
    except ModelHealthError as exc:
        sys.exit(f"Failed to stop recording: {exc}")
    print("Recording stopped. Videos are uploading.")

    # Wait for processing
    print("\nWaiting for upload and processing...")
    status = _poll_activity(service, activity)
    if status != ActivityStatus.ready:
        sys.exit(f"Activity did not reach ready state (status: {status}).")
    print(f"Activity is ready. ID: {activity.id}")
    print("\nDone. Run activity_analysis.py to analyze this activity.")


if __name__ == "__main__":
    args = docopt(__doc__)
    main(load_api_key(args["<api_key>"]))

File: examples/ts/activity_recording.ts
/**
 * Model Health TypeScript SDK — capture workflow.
 * Mirrors examples/python/activity_recording.py.
 *
 * Usage:
 *   npx tsx activity_recording.ts [<api_key>]
 */

import { ModelHealthService, ActivityType } from '@modelhealth/modelhealth';
import type {
  CheckerboardDetails, CheckerboardPlacement,
  CalibrationStatus, ActivityStatus, Analysis,
} from '@modelhealth/modelhealth';
import {
  loadApiKey, ANALYSIS_DATA_EXT,
  pickOne, confirm, prompt, saveFile, pollAnalysis, sleep, closePrompts,
} from './_shared.js';

const CB_WIDTH = 48;

function calibrationCallback(status: CalibrationStatus): void {
  let msg: string;
  switch (status.type) {
    case 'recording':
      msg = '  Recording...';
      break;
    case 'uploading':
      msg = `  Uploading (${status.uploaded}/${status.total} cameras)...`;
      break;
    case 'processing': {
      const pct = status.percent != null ? `${status.percent}%` : '--%';
      msg = `  Processing (${pct})...`;
      break;
    }
    case 'done':
      process.stdout.write('  Done.'.padEnd(CB_WIDTH) + '\n');
      return;
  }
  process.stdout.write(msg.padEnd(CB_WIDTH) + '\r');
}

interface CheckerboardPreset {
  rows: number | null;
  columns: number | null;
  squareSize: number | null;
  label: string;
}

const CHECKERBOARD_PRESETS: CheckerboardPreset[] = [
  { rows: 4, columns: 5, squareSize: 35, label: '4 × 5  —  35 mm squares  (standard A4)' },
  { rows: 4, columns: 5, squareSize: 50, label: '4 × 5  —  50 mm squares  (large A3)' },
  { rows: null, columns: null, squareSize: null, label: 'Other — enter manually' },
];

interface ActivityTypeOption {
  type: string | null;
  label: string;
}

const ACTIVITY_TYPE_OPTIONS: ActivityTypeOption[] = [
  { type: null,                               label: 'None — skip automatic analysis'  },
  { type: ActivityType.CounterMovementJump,   label: 'Counter Movement Jump'           },
  { type: ActivityType.Gait,                  label: 'Overground Walking'              },
  { type: ActivityType.TreadmillGait,         label: 'Treadmill Walking'               },
  { type: ActivityType.TreadmillRunning,      label: 'Treadmill Running'               },
  { type: ActivityType.OvergroundRunning,     label: 'Overground Running'              },
  { type: ActivityType.SitToStand,            label: 'Sit-to-Stand Transfer'           },
  { type: ActivityType.Squats,                label: 'Squats'                          },
  { type: ActivityType.RangeOfMotion,         label: 'Range of Motion'                 },
  { type: ActivityType.DropJump,              label: 'Drop Vertical Jump'              },
  { type: ActivityType.Hop,                   label: 'Hop Test'                        },
  { type: ActivityType.ChangeOfDirection,     label: '5-0-5 Test'                      },
  { type: ActivityType.Cut,                   label: 'Cutting Maneuver'                },
  { type: ActivityType.Sprint,                label: 'Sprint'                          },
  { type: ActivityType.LateralStepdown,       label: 'Lateral Step Down'               },
  { type: ActivityType.Lunge,                 label: 'Lunge'                           },
];

async function pollActivity(
  service: ModelHealthService,
  activity: Parameters<typeof service.activityStatus>[0]
): Promise<ActivityStatus> {
  while (true) {
    const status = await service.activityStatus(activity);
    if (status.type === 'uploading') {
      process.stdout.write(`  Uploading (${status.uploaded}/${status.total} cameras)...  \r`);
    } else if (status.type === 'processing') {
      process.stdout.write('  Processing...                              \r');
    } else {
      process.stdout.write('\n');
      return status;
    }
    await sleep(5_000);
  }
}

async function main() {
  const args = process.argv.slice(2);
  console.log('Connecting...');
  const service = new ModelHealthService({ apiKey: loadApiKey(args[0]), autoInit: false });
  await service.init();

  // Session
  console.log('\nCreating session...');
  const session = await service.createSession();
  console.log(`  Session ID: ${session.id}`);

  if (!session.qrcode) {
    console.error('Session has no QR code — cannot pair cameras.');
    process.exit(1);
  }

  const qrRes = await fetch(session.qrcode);
  const qrData = new Uint8Array(await qrRes.arrayBuffer());
  const qrPath = saveFile('qr-code.png', qrData);
  console.log(`  QR code saved to: ${qrPath}`);
  console.log('  Pair your cameras using the Model Health companion iOS app before continuing.');

  await prompt('\nPress Enter when cameras are ready...');

  // Camera calibration
  console.log('\nCheckerboard configuration:\n');
  const preset = await pickOne(CHECKERBOARD_PRESETS, 'Select checkerboard', p => p.label);

  let rows: number;
  let columns: number;
  let squareSize: number;

  if (preset.rows !== null && preset.columns !== null && preset.squareSize !== null) {
    rows = preset.rows;
    columns = preset.columns;
    squareSize = preset.squareSize;
  } else {
    rows      = parseInt((await prompt('  Internal rows: ')).trim(), 10);
    columns   = parseInt((await prompt('  Internal columns: ')).trim(), 10);
    squareSize = parseInt((await prompt('  Square size (mm): ')).trim(), 10);
  }

  console.log('\nCheckerboard placement:\n');
  const placement = await pickOne<CheckerboardPlacement>(
    ['perpendicular', 'parallel'],
    'Select placement',
    p => p === 'perpendicular' ? 'Perpendicular (upright, facing cameras)' : 'Parallel (flat on the floor)'
  );

  const checkerboard: CheckerboardDetails = { rows, columns, squareSize, placement };

  await prompt('\nPress Enter to start camera calibration...');
  console.log('Calibrating cameras...');
  await service.calibrateCamera(session, checkerboard, calibrationCallback);
  console.log('Camera calibration complete.');

  // Subject
  console.log('\nFetching subjects...');
  const subjects = await service.subjectList();

  let subject;
  if (subjects.length > 0 && await confirm(`Found ${subjects.length} subject(s). Select an existing one?`, true)) {
    console.log();
    subject = await pickOne(subjects, 'Select subject', s => `${s.name}  (ID ${s.id})`);
    console.log(`  Using: ${subject.name}`);
  } else {
    console.log('\nNew subject details:');
    const name   = (await prompt('  Name: ')).trim() || 'Anonymous';
    const weight = parseFloat((await prompt('  Weight (kg): ')).trim());
    const height = parseFloat((await prompt('  Height (cm): ')).trim());
    console.log('Creating subject...');
    subject = await service.createSubject({ name, weight, height });
    console.log(`  Subject created: ${subject.name} (ID ${subject.id})`);
  }

  // Subject calibration
  await prompt(`\nAsk ${subject.name} to stand in the neutral pose, then press Enter...`);
  console.log('Calibrating subject...');
  await service.calibrateSubject(subject, session, calibrationCallback);
  console.log('Subject calibration complete.');

  // Recording
  const activityName = (await prompt('\nActivity name (e.g. cmj, squat): ')).trim() || 'activity';

  console.log('\nAutomatic analysis (optional):\n');
  const selected = await pickOne(ACTIVITY_TYPE_OPTIONS, 'Select activity type', t => t.label);
  const activityTypeValue = selected.type;
  const activityTypeLabel = selected.label;

  await prompt(`\nAsk ${subject.name} to get ready, then press Enter to start recording...`);
  console.log('Recording...');
  const config = activityTypeValue ? { activityType: activityTypeValue as any } : undefined;
  const activity = await service.startRecording(activityName, session, config);
  console.log(`  Recording started (activity ${activity.id}).`);

  await prompt('\nPress Enter when the movement is complete to stop recording...');
  console.log('Stopping recording...');
  await service.stopRecording(session);
  console.log('Recording stopped. Videos are uploading.');

  // Wait for processing
  console.log('\nWaiting for upload and processing...');
  const finalStatus = await pollActivity(service, activity);

  let currentActivity = activity;

  if (finalStatus.type === 'analyzing') {
    const task: Analysis = { taskId: finalStatus.taskId };
    console.log(`Activity is ready. Automatic '${activityTypeLabel}' analysis has started.`);
    console.log('\nWaiting for analysis to complete...');
    const resultStatus = await pollAnalysis(service, task);

    if (resultStatus.type !== 'completed') {
      console.error(`Analysis did not complete (status: ${resultStatus.type}).`);
      process.exit(1);
    }
    console.log('Analysis complete.');

    const freshActivity = await service.fetchActivity(activity.id);
    currentActivity = freshActivity;
    const results = await service.analysisDataForActivity(freshActivity, ['report']);
    const slug = (freshActivity.name ?? freshActivity.id).replace(/ /g, '_');

    console.log('\nDownloading report...');
    for (const r of results) {
      const ext = ANALYSIS_DATA_EXT[r.type] ?? 'bin';
      const p = saveFile(`${slug}_${r.type}.${ext}`, r.data);
      console.log(`  Saved ${p}`);
    }
  } else if (finalStatus.type === 'ready') {
    console.log(`Activity is ready. ID: ${activity.id}`);
    console.log('Run activity_analysis.ts to analyze this activity.');
  } else {
    console.error(`Activity did not reach ready state (status: ${finalStatus.type}).`);
    process.exit(1);
  }

  // Update activity metadata (optional)
  console.log('\nUpdate activity (optional):');
  const currentTags = currentActivity.tags?.length ? currentActivity.tags.join(', ') : '(none)';
  console.log(`  Current tags: ${currentTags}`);
  const newName = (await prompt(`  New name (press Enter to keep '${currentActivity.name ?? currentActivity.id}'): `)).trim() || undefined;
  const addInput = (await prompt('  Tags to add, comma-separated (press Enter to skip): ')).trim();
  const addTags = addInput ? addInput.split(',').map(t => t.trim()).filter(Boolean) : [];
  const removeInput = (await prompt('  Tags to remove, comma-separated (press Enter to skip): ')).trim();
  const removeTags = removeInput ? removeInput.split(',').map(t => t.trim()).filter(Boolean) : [];
  if (newName || addTags.length || removeTags.length) {
    console.log('Updating activity...');
    try {
      const updated = await service.updateActivity(currentActivity, { name: newName, addTags, removeTags });
      console.log(`  Updated: ${updated.name ?? updated.id}`);
    } catch (err: any) {
      console.error(`Failed to update activity: ${err.message ?? err}`);
    }
  }
  console.log('\nDone.');
}

main().catch(err => { console.error(`Error: ${err.message ?? err}`); process.exit(1); })
  .finally(closePrompts);
